### PR TITLE
Add aqua, remove pirated methods

### PR DESCRIPTION
**Pull request by ReubenJ**
These methods were picked up by Aqua as being pirated. They should be defined in `HerbCore`. See https://github.com/Herb-AI/HerbCore.jl/pull/31. Removing these methods from this package means this is also a breaking change.

Aqua tests are now included in CI so that this won't happen by accident in the future.

The reason for changing the `RuleNode(..., grammar)` and `UniformHole(..., grammar)` constructors to functions was also for type-piracy reasons--defining a new constructor in this package (`HerbGrammar`) for a type defined in another package (`HerbCore`). They could have been moved to `HerbCore` as constructors, but then testing the constructors becomes difficult because they need a `ContextSensitiveGrammar` to test them, and that is defined in this package. So, in the end, it was just easier to change them to functions.

### CODE REVIEW

**Comment by ReubenJ** [comment]
Need to add a test for `grammar2symboltable`, but it looks like it's not straightforward to test. Variables defined in the test local scope don't seem to be picked up by the scope of `grammar2symboltable` when it tries to construct the table.

**Review by pwochner — APPROVED** [review]
Looks good. I like the examples that you added in the docstrings.

File: src/rulenode_operators.jl [hunk 1]
```diff
@@ -1,7 +1,71 @@
-HerbCore.RuleNode(ind::Int, grammar::AbstractGrammar) = RuleNode(ind, nothing, [Hole(get_domain(grammar, type)) for type ∈ grammar.childtypes[ind]])
-HerbCore.RuleNode(ind::Int, _val::Any, grammar::AbstractGrammar) = RuleNode(ind, _val, [Hole(get_domain(grammar, type)) for type ∈ grammar.childtypes[ind]])
+"""
+    holes_from_child_types(index::Integer, grammar::ContextSensitiveGrammar)
+
+Given the `index` of a rule in a `grammar`, create a vector of [`Hole`](@ref)s
+corresponding to the children of the rule at `index`.
+"""
+function holes_from_child_types(index::Integer, grammar::ContextSensitiveGrammar)
```
Comment (ReubenJ) [inline src/rulenode_operators.jl:7]: The name/purpose of this function could be clearer, as pointed out by @THinnerichs

**Review by ReubenJ — COMMENTED** [review]
Will address tomorrow

**Comment by ReubenJ** [comment]
Just noticed we can also drop `TreeView` and `DataStructures` as dependencies.

This brings the `using HerbGrammar` time down a bit.

<details><summary>Before</summary>
<p>

```julia
julia> @time_imports using HerbGrammar
      5.8 ms  MacroTools
      0.6 ms  SimpleTraits
      5.2 ms  Preferences
      0.4 ms  PrecompileTools
      0.6 ms  StaticArraysCore
    101.5 ms  StaticArrays
      1.1 ms  ArnoldiMethod
      0.7 ms  Statistics
      0.3 ms  StaticArrays → StaticArraysStatisticsExt
      0.5 ms  Compat
      0.5 ms  Inflate
      2.0 ms  OrderedCollections
     12.8 ms  DataStructures
      0.4 ms  Compat → CompatLinearAlgebraExt
               ┌ 9.7 ms SuiteSparse_jll.__init__()
     11.7 ms  SuiteSparse_jll
      0.9 ms  Serialization
               ┌ 6.6 ms SparseArrays.CHOLMOD.__init__() 57.96% compilation time
    116.0 ms  SparseArrays 3.27% compilation time
      0.5 ms  Statistics → SparseArraysExt
               ┌ 0.0 ms Distributed.__init__()
     16.6 ms  Distributed 53.83% compilation time
      0.7 ms  Mmap
      2.0 ms  SharedArrays
      8.2 ms  Graphs
      0.6 ms  LaTeXStrings
      0.6 ms  JLLWrappers
               ┌ 2.8 ms Bzip2_jll.__init__() 54.90% compilation time
      3.4 ms  Bzip2_jll 44.12% compilation time
               ┌ 1.3 ms Zlib_jll.__init__()
      3.4 ms  Zlib_jll
               ┌ 1.5 ms FreeType2_jll.__init__()
      2.1 ms  FreeType2_jll
      0.4 ms  Libuuid_jll
               ┌ 0.9 ms Expat_jll.__init__()
      1.5 ms  Expat_jll
               ┌ 7.9 ms Fontconfig_jll.__init__() 80.06% compilation time
      8.6 ms  Fontconfig_jll 72.95% compilation time
               ┌ 1.1 ms Graphite2_jll.__init__()
      1.8 ms  Graphite2_jll
               ┌ 2.0 ms Libiconv_jll.__init__()
      2.6 ms  Libiconv_jll
               ┌ 1.1 ms Libffi_jll.__init__()
      1.7 ms  Libffi_jll
               ┌ 11.6 ms CompilerSupportLibraries_jll.__init__()
     14.4 ms  CompilerSupportLibraries_jll
               ┌ 1.6 ms XML2_jll.__init__()
      2.3 ms  XML2_jll
               ┌ 2.4 ms Gettext_jll.__init__()
      3.1 ms  Gettext_jll
               ┌ 1.3 ms PCRE2_jll.__init__()
      3.1 ms  PCRE2_jll
               ┌ 10.3 ms Glib_jll.__init__()
     11.0 ms  Glib_jll
               ┌ 2.0 ms LLVMOpenMP_jll.__init__()
      2.7 ms  LLVMOpenMP_jll
               ┌ 2.0 ms Pixman_jll.__init__()
      2.6 ms  Pixman_jll
               ┌ 1.1 ms libpng_jll.__init__()
      1.7 ms  libpng_jll
               ┌ 3.4 ms Cairo_jll.__init__()
      4.1 ms  Cairo_jll
               ┌ 3.6 ms HarfBuzz_jll.__init__()
      4.3 ms  HarfBuzz_jll
               ┌ 6.2 ms ICU_jll.__init__()
      7.0 ms  ICU_jll
               ┌ 1.1 ms HarfBuzz_ICU_jll.__init__()
      1.9 ms  HarfBuzz_ICU_jll
               ┌ 4.4 ms OpenSSL_jll.__init__()
      5.2 ms  OpenSSL_jll
               ┌ 0.0 ms tectonic_jll.__init__()
      0.8 ms  tectonic_jll
               ┌ 0.0 ms Requires.__init__()
      0.7 ms  Requires
               ┌ 2.7 ms JpegTurbo_jll.__init__()
      3.5 ms  JpegTurbo_jll
               ┌ 1.2 ms LERC_jll.__init__()
      2.0 ms  LERC_jll
               ┌ 1.3 ms Zstd_jll.__init__()
      2.1 ms  Zstd_jll
               ┌ 1.5 ms Libtiff_jll.__init__()
      2.2 ms  Libtiff_jll
               ┌ 1.5 ms LittleCMS_jll.__init__()
      2.3 ms  LittleCMS_jll
               ┌ 1.5 ms OpenJpeg_jll.__init__()
      2.4 ms  OpenJpeg_jll
               ┌ 6.3 ms Poppler_jll.__init__()
      7.3 ms  Poppler_jll
               ┌ 63.2 ms TikzPictures.__init__() 42.66% compilation time
     67.6 ms  TikzPictures 39.89% compilation time
      1.0 ms  TikzGraphs
      0.8 ms  CommonSubexpressions
      0.8 ms  TreeView
      5.4 ms  AbstractTrees
      1.1 ms  HerbCore
    928.5 ms  HerbGrammar
```

</p>
</details> 

<details><summary>After</summary>
<p>

```julia
julia> @time_imports using HerbGrammar
      2.9 ms  Serialization
      2.1 ms  HerbCore
    924.6 ms  HerbGrammar
```

</p>
</details>

File: src/rulenode_operators.jl (hunk 1 above)
Comment (ReubenJ) [reply src/rulenode_operators.jl]: Dropping these functions instead as their utility is not that clear anyway, and we are not currently using them anywhere.

**Pull request merged**